File: Nissan-vercel-main/apps/api/app/routers/marketing_strategy.py
"""Marketing Strategy Advisor API — a Context Planner sub-module.

Stateless: given a context_id, reads that context's most recent analysis
(company summary + generated report) read-only and returns a Groq-generated
(deterministic-fallback) list of growth/marketing strategies. One bounded LLM
call, so POST is synchronous — no polling, no persistence.
"""
from __future__ import annotations

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from agents.marketing_strategy.service import ContextNotFound, suggest_strategies

router = APIRouter()


class SuggestRequest(BaseModel):
    tenant_id: str
    context_id: str


class Strategy(BaseModel):
    title: str
    category: str
    description: str
    reason: str
    expected_impact: str
    priority: str


class SuggestResponse(BaseModel):
    context_id: str
    status: str
    engine: str | None = None
    strategies: list[Strategy] = []
    errors: list[str] = []


@router.post("/suggest", response_model=SuggestResponse)
async def suggest_endpoint(body: SuggestRequest) -> SuggestResponse:
    try:
        result = await suggest_strategies(body.tenant_id, body.context_id)
    except ContextNotFound as e:
        raise HTTPException(status_code=404, detail=str(e))
    return SuggestResponse(**result)

File: Nissan-vercel-main/apps/web/src/lib/marketing-strategy.ts
/**
 * Marketing Strategy Advisor API client — Context Planner sub-module.
 * Stateless: POST a context_id, get back Groq-generated growth strategies.
 */
import { createServerFn } from '@tanstack/react-start'
import { getSupabaseServerClient } from './supabase.server'

const AGENT_BASE = (import.meta.env.VITE_AGENT_API_URL as string | undefined) ?? 'http://localhost:8000'

async function tenantId(supabase: ReturnType<typeof getSupabaseServerClient>) {
  const {
    data: { user },
  } = await supabase.auth.getUser()
  if (!user) throw new Error('Not authenticated')
  const { data } = await supabase.from('users').select('tenant_id').eq('id', user.id).single()
  if (!data?.tenant_id) throw new Error('User has no tenant')
  return data.tenant_id as string
}

export type StrategyPriority = 'high' | 'medium' | 'low'

export interface MarketingStrategy {
  title: string
  category: string
  description: string
  reason: string
  expected_impact: string
  priority: StrategyPriority
}

export interface StrategyResult {
  context_id: string
  status: 'ready' | 'no_analysis'
  engine: 'groq' | 'deterministic' | null
  strategies: Array<MarketingStrategy>
  errors: Array<string>
}

export const suggestStrategies = createServerFn({ method: 'POST' })
  .validator((d: { context_id: string }) => d)
  .handler(async ({ data }): Promise<StrategyResult> => {
    const supabase = getSupabaseServerClient()
    const tenant_id = await tenantId(supabase)

    const res = await fetch(`${AGENT_BASE}/marketing-strategy/suggest`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ tenant_id, context_id: data.context_id }),
    })
    if (!res.ok) {
      const err = (await res.json().catch(() => ({ detail: 'Failed to generate strategies' }))) as { detail?: string }
      throw new Error(typeof err.detail === 'string' ? err.detail : 'Failed to generate strategies')
    }
    return res.json() as Promise<StrategyResult>
  })
